src/core/algorithms/od/fastod/data_frame.cpp

```diff
@@ -0,0 +1,200 @@
+
+    RecognizeAttributesWithRanges();
+}
+
+int DataFrame::GetValue(int tuple_index, AttributeSet::size_type attribute_index) const {
+    return data_[attribute_index][tuple_index];
+}
+
+std::vector<std::vector<DataFrame::value_indexes_t>> const& DataFrame::GetDataRanges() const {
+    return data_ranges_;
+}
+
+size_t DataFrame::GetRangeIndexByItem(size_t item, AttributeSet::size_type attribute) const {
+    return range_item_placement_[attribute][item];
+}
+
+AttributeSet::size_type DataFrame::GetColumnCount() const {
+    return data_.size();
+}
+
+std::size_t DataFrame::GetTupleCount() const {
+    return data_.size() > 0 ? data_[0].size() : 0;
+}
+
+bool DataFrame::IsAttributesMostlyRangeBased(AttributeSet attributes) const {
+    if (!attributes.any()) {
+        return false;
+    }
+
+    AttributeSet remaining_attrs = intersect(attrs_with_ranges_, attributes);
+
+    AttributeSet::size_type attrs_count = attributes.count();
+    AttributeSet::size_type remaining_attrs_count = remaining_attrs.count();
+
+    double constexpr accept_range_based_partition_factor = 0.5;
+
+    return static_cast<double>(remaining_attrs_count) / attrs_count >=
+           accept_range_based_partition_factor;
+}
+
+DataFrame DataFrame::FromCsv(std::filesystem::path const& path, char separator, bool has_header,
+                             config::EqNullsType is_null_equal_null) {
+    std::shared_ptr<CSVParser> parser = std::make_shared<CSVParser>(path, separator, has_header);
+    return FromInputTable(parser, is_null_equal_null);
+}
+
+DataFrame DataFrame::FromInputTable(config::InputTable input_table,
+                                    config::EqNullsType is_null_equal_null) {
+    std::vector<model::TypedColumnData> columns_data =
+            model::CreateTypedColumnData(*input_table, is_null_equal_null);
+
+    return DataFrame(std::move(columns_data));
+}
+
+void DataFrame::RecognizeAttributesWithRanges() {
+    double constexpr accept_factor = 0.001;
+
+    for (size_t i = 0; i < data_ranges_.size(); ++i) {
+        const size_t items_count = data_[i].size();
+        const size_t ranges_count = data_ranges_[i].size();
+
+        if (static_cast<double>(ranges_count) / items_count >= accept_factor) {
+            attrs_with_ranges_.set(i, true);
+        }
+    }
+}
+
+std::vector<std::pair<std::byte const*, int>> DataFrame::CreateIndexedColumnData(
+        model::TypedColumnData const& column) {
+    const std::vector<std::byte const*> data = column.GetData();
+    std::vector<std::pair<std::byte const*, int>> indexed_column_data(data.size());
+
+    for (std::size_t i = 0; i < data.size(); ++i) {
+        indexed_column_data[i] = std::make_pair(data[i], i);
+    }
+
+    return indexed_column_data;
+}
+
+std::vector<int> DataFrame::ConvertColumnDataToIntegers(model::TypedColumnData const& column) {
+    using data_t = std::pair<std::byte const*, int>;
+
+    std::function<bool(data_t, data_t)> less_mixed = [&column](data_t l, data_t r) {
```

Comment:
Using `std::function` looks looks bad enough. The `std::function` is needed for type erasure. It does provide a single interface for different operations, but it is not free. Because of this, runtime overhead is added, for which you have to pay with execution time.

Try to avoid using it, for example by using templates.